openforcefield/topology/molecule.py

```diff
@@ -3176,6 +3376,343 @@ def from_openeye(oemol, allow_undefined_stereo=False):
         return toolkit.from_openeye(
             oemol, allow_undefined_stereo=allow_undefined_stereo)
 
+    def to_qcschema(self, multiplicity=1, conformer=0):
+        """
+        Generate the qschema input format used to submit jobs to archive
+        or run qcengine calculations locally, the molecule is placed in canonical order first.
+        spec can be found here <https://molssi-qc-schema.readthedocs.io/en/latest/index.html>
+
+        .. warning :: This API is experimental and subject to change.
+
+        Parameters
+        ----------
+        multiplicity : int, default=1, the multiplicity of the molecule required for qcschema
+
+        conformer : int, default=0, the index of the conformer that should be used for qcschema
+
+        Returns
+        ---------
+        qcelemental.models.Molecule : a validated qcschema
+
+        Example
+        -------
+
+        Create and validate a qcelemental input
+
+        >>> import qcelemental as qcel
+        >>> mol = Molecule.from_smiles('CC')
+        >>> mol.generate_conformers(n_conformers=1)
+        >>> qcschema = mol.to_qcschema()
+
+        Raises
+        --------
+        ImportError : if qcelemental is not installed; the qcschema can not be validated.
+
+        InvalidConformerError : if there is no conformer found at the given index.
+        """
+
+        try:
+            import qcelemental as qcel
+        except ImportError:
+            raise ImportError('Please install QCElemental via conda install -c conda-forge qcelemental '
+                              'to validate the schema')
+
+        # get a canonical ordered version of the molecule
+        canonical_mol = self.canonical_order_atoms()
+
+        # get/ check the geometry
+        try:
+            geometry = canonical_mol.conformers[conformer].in_units_of(unit.bohr)
+        except (IndexError, TypeError):
+            raise InvalidConformerError('The molecule must have a conformation to produce a valid qcschema; '
+                                        f'no conformer was found at index {conformer}.')
+
+        # Gather the required qschema data
+        charge = sum([atom.formal_charge for atom in canonical_mol.atoms])
+        connectivity = [(bond.atom1_index, bond.atom2_index, bond.bond_order) for bond in canonical_mol.bonds]
+        symbols = [Element.getByAtomicNumber(atom.atomic_number).symbol for atom in canonical_mol.atoms]
+
+        schema_dict = {'symbols': symbols, 'geometry': geometry, 'connectivity': connectivity,
+                       'molecular_charge': charge, 'molecular_multiplicity': multiplicity}
+
+        return qcel.models.Molecule.from_data(schema_dict, validate=True)
+
+    @classmethod
+    def from_mapped_smiles(cls, mapped_smiles, toolkit_registry=GLOBAL_TOOLKIT_REGISTRY, allow_undefined_stereo=False):
```

Comment:
(Blocking) Let's make this private (`def _from_mapped_smiles`) until we can discuss this more, for two reasons:
1) #412 will make this break for a lot of molecules
2) I think that supporting `offmol._properties['atom_map']` leads down a dangerous road, and would like to chat in person before we commit to implementing that.

Reply:
Yeah, it might be better to follow the toolkits and have an atom map index attribute on each atom which can be set? This could be a safer way of storing the atom map and reordering molecules using the indexes? This would also be helpful for the fragmenter move I think.

I think #412 is fixed using the atom map stashing that I put into the `from_smiles` method in toolkit but I should possibly test it on more molecules?